src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1457,7 +1457,7 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
                 var viableEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
                 foreach (var candidate in entryPointCandidates)
                 {
-                    if (!candidate.HasEntryPointSignature(this))
+                    if (!candidate.HasEntryPointSignature(this, diagnostics))
```

Comment:
Tests for all of the permutations of main method declarations are going in the next PR.